lib/collection/src/shards/shard_holder/resharding.rs

```diff
@@ -364,6 +369,107 @@ impl ShardHolder {
         Ok(())
     }
 
+    pub fn split_by_mode(
+        &self,
+        shard_id: ShardId,
+        operation: CollectionUpdateOperations,
+    ) -> OperationsByMode {
+        let Some(state) = self.resharding_state() else {
+            return operation.into();
+        };
+
+        // Resharding *UP*
+        // ┌────────────┐   ┌──────────┐
+        // │            │   │          │
+        // │ Shard 1    │   │ Shard 2  │
+        // │ Non-Target ├──►│ Target   │
+        // │ Sender     │   │ Receiver │
+        // │            │   │          │
+        // └────────────┘   └──────────┘
+        //
+        // Resharding *DOWN*
+        // ┌────────────┐   ┌──────────┐
+        // │            │   │          │
+        // │ Shard 1    │   │ Shard 2  │
+        // │ Non-Target │◄──┤ Target   │
+        // │ Receiver   │   │ Sender   │
+        // │            │   │          │
+        // └────────────┘   └──────────┘
+
+        // Target shard of the resharding operation. This is the shard that:
+        //
+        // - *created* during resharding *up*
+        // - *deleted* during resharding *down*
+        let is_target_shard = shard_id == state.shard_id;
+
+        // Shard that will be *receiving* migrated points during resharding:
+        //
+        // - *target* shard during resharding *up*
+        // - *non* target shards during resharding *down*
+        let is_receiver_shard = match state.direction {
+            ReshardingDirection::Up => is_target_shard,
+            ReshardingDirection::Down => !is_target_shard,
+        };
+
+        // Shard that will be *sending* migrated points during resharding:
+        //
+        // - *non* target shards during resharding *up*
+        // - *target* shard during resharding *down*
+        let is_sender_shard = !is_receiver_shard;
+
+        // We split update operations:
+        //
+        // - on *receiver* shards during `MigratingPoints` stage (for all operations except `upsert`)
+        // - and on *sender* shards during `ReadHashRingCommitted` stage when resharding *up*
+
+        let should_split_receiver = is_receiver_shard
+            && state.stage == ReshardStage::MigratingPoints
+            && !operation.is_upsert_points();
+
+        let should_split_sender = is_sender_shard
+            && state.stage >= ReshardStage::ReadHashRingCommitted
+            && state.direction == ReshardingDirection::Up;
+
+        if !should_split_receiver && !should_split_sender {
+            return operation.into();
+        }
+
+        // There's no point splitting delete operations
+        if operation.is_delete_points() {
+            return operation.into();
+        }
```

Comment:
Do we return an error if deletion fails? If so, I think we should split it to prevent this.

Reply:
It does not error.